File: scripts/validate_weapons.py
import re
import json
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DB_PATH = ROOT / 'weaponDatabase.js'
OUT_DIR = ROOT / 'reports'
OUT_DIR.mkdir(exist_ok=True)
OUT_PATH = OUT_DIR / 'weapon-audit.json'

with open(DB_PATH, 'r', encoding='utf8') as f:
    src = f.read()

def find_array_block(src, key):
    # find the start of '<key>:' then the following '[' and return the content until matching ']'
    m = re.search(rf"\b{key}\s*:\s*\[", src)
    if not m:
        return None
    i = m.end()  # position after '['
    depth = 1
    start = i
    while i < len(src):
        c = src[i]
        if c == '[':
            depth += 1
        elif c == ']':
            depth -= 1
            if depth == 0:
                return src[start:i]
        i += 1
    return None

def extract_objects(array_text):
    objs = []
    i = 0
    while i < len(array_text):
        # find next '{'
        m = re.search(r"\{", array_text[i:])
        if not m:
            break
        j = i + m.start()
        depth = 0
        k = j
        while k < len(array_text):
            if array_text[k] == '{':
                depth += 1
            elif array_text[k] == '}':
                depth -= 1
                if depth == 0:
                    objs.append(array_text[j:k+1])
                    i = k+1
                    break
            k += 1
        else:
            break
    return objs

def normalize_name(name):
    return re.sub(r'[^a-z0-9]', '', name.lower())

canonical_warbonds = set([
  'free','helldivers-mobilize','steeled-veterans','cutting-edge','democratic-detonation',
  'polar-patriots','viper-commandos','freedom-flame','chemical-agents','truth-enforcers',
  'urban-legends','servants-freedom','borderline-justice','masters-ceremony','force-of-law',
  'control-group','odst','dust-devils'
])

report = {
  'totalItems': 0,
  'invalidWarbonds': [],
  'duplicates': [],
  'crossCategoryDuplicates': [],
  'placeholders': [],
  'defaultGrantedCandidates': [],
  'primariesMissingType': [],
  'itemsByCategoryCount': {}
}

seen = {}
name_to_categories = {}

for category in ('primary','secondary','grenade','stratagem'):
    block = find_array_block(src, category)
    if block is None:
        report['itemsByCategoryCount'][category] = 0
        continue
    objs = extract_objects(block)
    report['itemsByCategoryCount'][category] = len(objs)
    for obj in objs:
        # extract name, warbond, type
        name_m = re.search(r"name:\s*\"([^\"]+)\"|name:\s*'([^']+)'", obj)
        name = (name_m.group(1) or name_m.group(2)) if name_m else '(no-name)'
        warbond_m = re.search(r"warbond:\s*\"([^\"]+)\"|warbond:\s*'([^']+)'", obj)
        warbond = (warbond_m.group(1) or warbond_m.group(2)) if warbond_m else 'UNKNOWN'
        type_m = re.search(r"type:\s*\"([^\"]+)\"|type:\s*'([^']+)'", obj)
        typ = (type_m.group(1) or type_m.group(2)) if type_m else None

        report['totalItems'] += 1

        if warbond not in canonical_warbonds:
            report['invalidWarbonds'].append({'category': category, 'name': name, 'warbond': warbond})

        if category == 'primary' and not typ:
            report['primariesMissingType'].append({'name': name, 'warbond': warbond})

        key = f"{normalize_name(name)}|{category}|{warbond}"
        seen.setdefault(key, []).append({'name': name, 'category': category, 'warbond': warbond})

        norm = normalize_name(name)
        name_to_categories.setdefault(norm, set()).add(category)

        # Only mark explicit placeholder items by name. Previously we treated
        # entire warbonds like 'control-group' and 'dust-devils' as placeholders
        # during early data-entry; now the data file contains real entries and
        # we should only flag items that explicitly contain 'placeholder'.
        if re.search(r'placeholder', name, re.I):
            report['placeholders'].append({'category': category, 'name': name, 'warbond': warbond})

        if category == 'stratagem' and re.match(r'^(resupply|reinforce|hellbomb)$', name, re.I):
            report['defaultGrantedCandidates'].append({'name': name, 'category': category, 'warbond': warbond})

for k, v in seen.items():
    if len(v) > 1:
        report['duplicates'].append({'key': k, 'count': len(v), 'examples': v[:5]})

for norm, cats in name_to_categories.items():
    if len(cats) > 1:
        report['crossCategoryDuplicates'].append({'nameNorm': norm, 'categories': list(cats)})

with open(OUT_PATH, 'w', encoding='utf8') as f:
    json.dump(report, f, indent=2)

print('Weapon database audit complete.')
print('Total items scanned:', report['totalItems'])
print('Invalid warbonds:', len(report['invalidWarbonds']))
print('Duplicates:', len(report['duplicates']))
print('Cross-category duplicates:', len(report['crossCategoryDuplicates']))
print('Placeholders detected:', len(report['placeholders']))
print('Default-granted stratagem candidates:', len(report['defaultGrantedCandidates']))
print('Primaries missing type:', len(report['primariesMissingType']))
print('Full JSON report written to:', OUT_PATH)

exit_code = 2 if (report['invalidWarbonds'] or report['duplicates'] or report['crossCategoryDuplicates']) else 0
raise SystemExit(exit_code)

File: scripts/validateWeapons.js
const fs = require('fs');
const path = require('path');

const dbPath = path.join(__dirname, '..', 'weaponDatabase.js');
const weapons = require(dbPath);

// Canonical/known warbond slugs (allow 'free' alias used in DB)
const canonicalWarbonds = new Set([
  'free','helldivers-mobilize','steeled-veterans','cutting-edge','democratic-detonation',
  'polar-patriots','viper-commandos','freedom-flame','chemical-agents','truth-enforcers',
  'urban-legends','servants-freedom','borderline-justice','masters-ceremony','force-of-law',
  'control-group','odst','dust-devils'
]);

function normalizeName(name) {
  return String(name).toLowerCase().replace(/[^a-z0-9]/g, '');
}

const report = {
  totalItems: 0,
  invalidWarbonds: [],
  duplicates: [],
  crossCategoryDuplicates: [],
  placeholders: [],
  defaultGrantedCandidates: [],
  primariesMissingType: [],
  itemsByCategoryCount: {}
};

const seen = new Map(); // key -> {count, examples}
const nameToCategories = new Map();

Object.keys(weapons).forEach(category => {
  const arr = weapons[category] || [];
  report.itemsByCategoryCount[category] = arr.length;
  arr.forEach(item => {
    report.totalItems += 1;
    const warbond = item.warbond || 'UNKNOWN';
    const name = item.name || '(no-name)';

    // invalid warbond
    if (!canonicalWarbonds.has(warbond)) {
      report.invalidWarbonds.push({ category, name, warbond });
    }

    // primary missing type
    if (category === 'primary' && !item.type) {
      report.primariesMissingType.push({ name, warbond });
    }

    // duplicates detection (within same category + warbond)
    const key = `${normalizeName(name)}|${category}|${warbond}`;
    if (!seen.has(key)) seen.set(key, []);
    seen.get(key).push({ name, category, warbond });

    // cross-category mapping by normalized name
    const norm = normalizeName(name);
    if (!nameToCategories.has(norm)) nameToCategories.set(norm, new Set());
    nameToCategories.get(norm).add(category);

    // placeholders
    if (/placeholder/i.test(name) || ['control-group','dust-devils'].includes(warbond)) {
      report.placeholders.push({ category, name, warbond });
    }

    // default-granted stratagems candidates
    if (category === 'stratagem') {
      if (/^resupply$/i.test(name) || /^reinforce$/i.test(name) || /^hellbomb$/i.test(name)) {
        report.defaultGrantedCandidates.push({ name, category, warbond });
      }
    }
  });
});

// collect duplicates
for (const [key, items] of seen.entries()) {
  if (items.length > 1) {
    report.duplicates.push({ key, count: items.length, examples: items.slice(0,5) });
  }
}

// cross-category duplicates (same normalized name across multiple categories)
for (const [norm, categories] of nameToCategories.entries()) {
  if (categories.size > 1) {
    report.crossCategoryDuplicates.push({ nameNorm: norm, categories: Array.from(categories) });
  }
}

// Write report
const outDir = path.join(__dirname, '..', 'reports');
if (!fs.existsSync(outDir)) fs.mkdirSync(outDir);
const outPath = path.join(outDir, 'weapon-audit.json');
fs.writeFileSync(outPath, JSON.stringify(report, null, 2), 'utf8');

// Console summary
console.log('Weapon database audit complete.');
console.log(`Total items scanned: ${report.totalItems}`);
console.log(`Invalid warbonds: ${report.invalidWarbonds.length}`);
console.log(`Duplicates (same name+category+warbond): ${report.duplicates.length}`);
console.log(`Cross-category duplicates: ${report.crossCategoryDuplicates.length}`);
console.log(`Placeholders detected: ${report.placeholders.length}`);
console.log(`Default-granted stratagem candidates: ${report.defaultGrantedCandidates.length}`);
console.log(`Primaries missing type: ${report.primariesMissingType.length}`);
console.log(`Full JSON report written to: ${outPath}`);

// Exit with non-zero if critical issues found (duplicates or invalid warbonds)
if (report.invalidWarbonds.length > 0 || report.duplicates.length > 0 || report.crossCategoryDuplicates.length > 0) {
  process.exitCode = 2;
}
